File: linked-lists/3.c
/**
 * 3.c
 * Linked Lists
 * https://github.com/ashenm/dsa
 *
 * Implements an abstract manipulation method
 *
 * Ashen Gunaratne
 * mail@ashenm.ml
 *
 */

#include <stdio.h>
#include <stdlib.h>

typedef struct node
{
    int value;
    struct node *next;
} node;

// conjoins a new node
node* append (node *head, int value);

int main (void)
{

    node *head = NULL;

    head = append(head, 25);
    head = append(head, 50);

    for (node *elect = head; elect != NULL; elect = elect -> next)
    {
        printf("%i\n", elect -> value);
    }

    // BUG
    // memory leak
    free(head);

}

node* append (node *head, int value)
{

    node *inset = malloc(sizeof(node));

    if (inset == NULL)
    {
        return NULL;
    }

    inset -> value = value;
    inset -> next = NULL;

    if (head == NULL)
    {
        return inset;
    }

    node *elect = head;

    while (elect -> next != NULL)
    {
        elect = elect -> next;
    }

    elect -> next = inset;

    return head;

}

// vim: set expandtab shiftwidth=4:

File: linked-lists/3.js
/**
 * 3.js
 * Linked Lists
 * https://github.com/ashenm/dsa
 *
 * Implements an abstract manipulation method
 *
 * Ashen Gunaratne
 * mail@ashenm.ml
 *
 */

'use strict';

class Node {

  constructor (value, next=null) {
    this.value = value;
    this.next = next;
  }

}

const main = function () {

  let head = null;

  head = append(head, 25);
  head = append(head, 50);

  for (let elect = head; elect !== null; elect = elect.next) {
    console.log(elect.value);
  }

};

const append = function (head, value) {

  const inset = new Node(value);

  if (!head) {
    return inset;
  }

  let elect = head;

  while (elect.next !== null) {
    elect = elect.next;
  }

  elect.next = inset;

  return head;

};

if (require.main === module) {
  main();
}

// vim: set expandtab shiftwidth=2:
